### PR TITLE
Feature/1 fuzz tests for swap optimistic

### PR DESCRIPTION
**Pull request by truemike**

### CODE REVIEW

File: test/UniswapV2PairTest.sol [hunk 1]
```diff
@@ -145,6 +227,25 @@ contract UniswapV2PairTest is Test {
...
+
+    function min(uint a, uint b) private returns (uint) {
+        return a < b ? a : b;
+    }
+
+    function max(uint a, uint b) private returns (uint) {
+        return a > b ? a : b;
+    }
+
+    function mapSeedToRange(uint seed, uint minInclusive, uint maxInclusive) private returns (uint) {
```
Comment (truemike) [inline test/UniswapV2PairTest.sol]: I thought all of the +1's to make parameters exclusive for the max were confusing the calling logic, so I just made both parameters inclusive.

File: test/UniswapV2PairTest.sol [hunk 2]
```diff
@@ -104,38 +140,84 @@ contract UniswapV2PairTest is Test {
...
     }
 
-    function runOptimisticSwapTestCase(uint outputAmount, uint token0Amount, uint token1Amount, uint inputAmount) private {
+    function runOptimisticSwapTestCase(uint outputAmount, uint token0Amount, uint token1Amount) private {
         addLiquidity(token0Amount, token1Amount);
+        uint inputAmount = computeOptimisticInputAmount(outputAmount);
         token0.transfer(address(pair), inputAmount);
 
-        vm.expectRevert("UniswapV2: K");
+        // The reason we revert may change depending on the relative values of the output amount and available liquidity:
```
Comment (truemike) [inline test/UniswapV2PairTest.sol:260]: This was a corner case I just got to the bottom of in my last pass. I'd been thinking output and token0 values of 1 and 2 respectively were causing liquidity issues in the core logic, but it was just the reason we were reverting as expected was different (2 and 2 correctly declared INSUFFICIENT_LIQUIDITY, but all other cases off the border declared invalid K). This let me reduce the minimum token0 amount once I realized it...

File: test/UniswapV2PairTest.sol [hunk 3]
```diff
@@ -104,38 +140,84 @@ contract UniswapV2PairTest is Test {
...
+    function testOptimisticFuzzTestCase(uint outputVal, uint token0Val, uint token1Val) public {
+        (uint outputAmount, uint token0Amount, uint token1Amount) = generateOptimisticFuzzCase(outputVal, token0Val, token1Val);
+        runOptimisticSwapTestCase(outputAmount, token0Amount, token1Amount);
+    }
+
+    function generateOptimisticFuzzCase(uint outputAmountSeed, uint token0AmountSeed, uint token1AmountSeed) private
+            returns (uint _outputAmount, uint _token0Amount, uint _token1Amount) {
+
+        // Token 0 needs at least: A) one output amount to take out and B) one remaining, otherwise all removals will fail:
+        uint minToken0Amount = 2;
```
Comment (truemike) [inline test/UniswapV2PairTest.sol]: Expanding on the comment, all swaps require output values greater than one and the remaining liquidity for the token being at least one. With values of token0 less than two, it's impossible to execute a valid swap.

File: test/UniswapV2PairTest.sol [hunk 4]
```diff
@@ -104,38 +140,84 @@ contract UniswapV2PairTest is Test {
...
+        // Token 0 needs to leave at least 2 in the global supply for an input transfer that will be one or greater with fees (beyond one) taken out:
+        uint maxToken0Amount = MAXIMUM_SUPPLY - 2;
```
Comment (truemike) [inline test/UniswapV2PairTest.sol]: If token0 gets the full supply without leaving two behind, the input transfer before the swap is going to fail. The input amount must be at least two as removing fees from one (or less) drops the output amount below one which fails the swap.

File: test/UniswapV2PairTest.sol [hunk 5]
```diff
@@ -104,38 +140,84 @@ contract UniswapV2PairTest is Test {
...
+        _token0Amount = mapSeedToRange(token0AmountSeed, minToken0Amount, maxToken0Amount);
+
+        // The liquidity between the token amounts needs to be 1001 or greater or else mint will fail:
+        uint minToken1Amount = divCeil((MINIMUM_LIQUIDITY + 1) ** 2, _token0Amount);
+        _token1Amount = mapSeedToRange(token1AmountSeed, minToken1Amount, MAXIMUM_SUPPLY);
+
+        // We need to make sure there is enough liquidity to have one remaining, so max liquidity-based output is always one less than token amount:
+        uint maxOutputAmountLiquidity = _token0Amount - 1;
+        // We must also ensure enough tokens remain in global supply to allow the input transfer (when incorporating the 0.03% fee, rounding down):
+        uint maxOutputAmountTotalSupply = ((MAXIMUM_SUPPLY - _token0Amount) * 997) / 1000;
```
Comment (truemike) [inline test/UniswapV2PairTest.sol]: Makes sure that adding back fees (dividing output amount by 99.7%) does not force the input transfer to exceed remaining global supply not moved into the pair.

File: test/UniswapV2PairTest.sol [hunk 6]
```diff
@@ -13,7 +13,9 @@ contract UniswapV2PairTest is Test {
     IERC20 private token0;
     IERC20 private token1;
 
-    uint112 constant  MINIMUM_LIQUIDITY = 10 ** 3;
+    uint112 constant MINIMUM_LIQUIDITY = 10 ** 3;
+    uint112 constant MAXIMUM_SUPPLY = type(uint112).max ;
```
Comment (duelinggalois) [inline test/UniswapV2PairTest.sol]: I think there are two maxes here, `MAXIMUM_SUPPLY` could be renamed to `MAXIMUM_UNI_RESERVE` and then a max token variable could be created that would be passed into `getStubToken` and tests would still pass with this change. 
```
    uint256 constant MAX_TOKEN = type(uint256).max;
    uint112 constant MAXIMUM_UNI_RESERVE = type(uint112).max ;
```

File: test/UniswapV2PairTest.sol [hunk 7]
```diff
@@ -66,36 +68,70 @@ contract UniswapV2PairTest is Test {
...
+            returns (uint _swapAmount, uint _token0Amount, uint _token1Amount) {
+
+        // The swap amount cannot be zero, but others should be generally accepted:
+        uint minSwapAmount = 1;
+        // The swap amount cannot match/exceed token 0's amount (as it'd pull too much token 1 liquidity from the pool), so split max supply in half:
+        uint maxSwapAmount = MAXIMUM_SUPPLY/2; // +1 should fail as the swap input tranfer would fail
+        _swapAmount = mapSeedToRange(swapSeed, minSwapAmount, maxSwapAmount);
+
+        // Token 0 needs to be at least one more than the swap amount (otherwise it'd drain all liquidity):
+        uint minToken0Amount = _swapAmount + 1;
```
Comment (duelinggalois) [inline test/UniswapV2PairTest.sol]: I think the min here is actually the minimum liquidity enforced by uniswap of 1000. Its actually not possible to drain all the liquidity when using `x*y=k`. Since `x_0 and y_0 > 0` so is `k > 0`, and if `x_1 = 0` that would imply `k_1 = 0` but `k_0 > 0` and `k_0 = k_1` according to the constant product formula.

File: .vscode/launch.json [hunk 8]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (duelinggalois) [inline .vscode/launch.json]: This file is confusing me, Im not seeing it in the target branch, but he pr shows lines removed. I probably did something weird testing out IDEs that led to this, but I think we should take the file out of the pr.

**Comment by cirethic** [comment]
Can we remove the compiler warnings before merging into mainstream? 

```
Compiler run successful (with warnings)
warning[2018]: Warning: Function state mutability can be restricted to pure
   --> test/UniswapV2PairTest.sol:485:5:
    |
485 |     function divCeil(uint256 numerand, uint256 divisor) private returns (uint256) {
    |     ^ (Relevant source part starts here and spans across multiple lines).



warning[2018]: Warning: Function state mutability can be restricted to pure
   --> test/UniswapV2PairTest.sol:490:5:
    |
490 |     function min(uint256 a, uint256 b) private returns (uint256) {
    |     ^ (Relevant source part starts here and spans across multiple lines).



warning[2018]: Warning: Function state mutability can be restricted to pure
   --> test/UniswapV2PairTest.sol:494:5:
    |
494 |     function max(uint256 a, uint256 b) private returns (uint256) {
    |     ^ (Relevant source part starts here and spans across multiple lines).



warning[2018]: Warning: Function state mutability can be restricted to pure
   --> test/UniswapV2PairTest.sol:498:5:
    |
498 |     function mapSeedToRange(
    |     ^ (Relevant source part starts here and spans across multiple lines).


```

**Review by duelinggalois — APPROVED** [review]

**Review by cirethic — APPROVED** [review]
👍
